lib/hal/portable/st/aws_hal_perf_cntr.c

```diff
@@ -0,0 +1,233 @@
+/**
+ * @brief Enable interrupt for the timer.
+ */
+
+void HAL_TIM_Base_MspInit( TIM_HandleTypeDef * pxTimerHandle )
+{
+
+    if ( TIM5 == pxTimerHandle->Instance )
+    {
+        /* Enable timer clock. */
+        __HAL_RCC_TIM5_CLK_ENABLE();
+
+        /* Set interrupt priority, and enable counter overflow interrupt. */
+        HAL_NVIC_SetPriority(TIM5_IRQn, configHAL_PERF_COUNTER_INTERRUPT_PRIORITY, 0 );
+        HAL_NVIC_EnableIRQ(TIM5_IRQn);
+    }
+}
+
+/*-----------------------------------------------------------*/
+/**
+ * @brief Clean up
+ */
+
+void HAL_TIM_Base_MspDeInit( TIM_HandleTypeDef * pxTimerHandle )
+{
+    if (TIM5 == pxTimerHandle->Instance )
+    {
+        /* Disable counter overflow interrupt. */
+        HAL_NVIC_DisableIRQ(TIM5_IRQn);
+
+        /* Stop TIM5 source clock. */
+        __HAL_RCC_TIM5_CLK_DISABLE();
+    }
+
+}
+
+/*-----------------------------------------------------------*/
+/**
+ * @brief Implement overflow ISR for TIM5
+ */
+void HAL_TIM5_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+    ullTimerOverflow++;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief TIM5 ISR
+ *
+ * @note a cleaner way to handle timer IRQ for ST is to implement a branch for
+ * .Instance == TIM5 in HAL_TIM_PeriodElapsedCallback(). Yet, to not expose static
+ * variable ullTimerOverflow, enabling USE_HAL_TIM_REGISTER_CALLBACKS and use
+ * .PeriodElapsedCallback for TIM5.
+ */
+
+void TIM5_IRQHandler(void)
+{
+    HAL_TIM_IRQHandler( &xTimerHandle );
+}
+/*-----------------------------------------------------------*/
+
+void aws_hal_perfcounter_open(void)
+{
+    uint32_t ulTimClock = 0;
+
+    /* Clear static variable which tracks total number of timer overflow. */
+    ullTimerOverflow = 0;
+
+    /* Get timer clock frequency. */
+    ulTimClock = HAL_RCC_GetPCLK1Freq();
+
+    /* Check user input frequency is valid. */
+    configASSERT( HAL_PERF_COUNTER_FREQ <= ulTimClock );
+
+    /* From STM32L475xx chip datasheet:
+     * - TIM2 and TIM5 are 32-bit general purpose timers, while others are all 16-bit.
+     * - Peripheral current consumption (Range 1/Range 2/Low power and sleep):
+     *   TIM2: 6.8/6.5/6.3 (uA/MHZ)
+     *   TIM5: 6.5/5.5/6.1 (uA/MHZ)
+     * To minimize the need for handling timer overflow interrupt with relative low cost,
+     * prefer to use TIM5 as the hardware timer for performance counter.
+     */
+    xTimerHandle.Instance = TIM5;
+
+    /* To minimize the need for serving overflow IRQ,
+     * .Period is always set to the max value the counter allows.
+     * Performance counter resolution is adjusted with .Prescaler.
+     */
+    xTimerHandle.Init.Prescaler = ulTimClock / HAL_PERF_COUNTER_FREQ;
+    xTimerHandle.Init.CounterMode = TIM_COUNTERMODE_UP;
+    xTimerHandle.Init.Period = HW_TIMER_32_CONST_PERIOD;
+    xTimerHandle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    xTimerHandle.Init.RepetitionCounter = 0x00;
```

Comment:
can you assign a variable instead of hardcoding 0x00

Reply:
Will make a macro in next PR.